casanovo/denovo/model.py

```diff
@@ -989,6 +989,190 @@ def configure_optimizers(
+        # Continually grab num_pairs items from all_psm until list is exhausted
+        while len(all_psm) > 0:
+            batch = all_psm[: self.num_pairs]
+            all_psm = all_psm[self.num_pairs :]
+            batch = list(zip(*batch))
+            encoded_ms = (
+                torch.stack([a[0] for a in batch[0]]),
+                torch.stack([a[1] for a in batch[0]]),
+            )
+            prec_data = torch.stack(batch[1])
+            pep_str = list(batch[2])
+            indexes = [a[1] for a in batch[3]]
+            t_or_ds = batch[4]
+            yield (indexes, t_or_ds, pep_str, prec_data, encoded_ms)
+
+    def on_predict_batch_end(
+        self,
+        outputs: List[Tuple[np.ndarray, List[str], torch.Tensor]],
+        *args,
+    ) -> None:
+        if self.out_writer is None:
+            return
+        for (
+            indexes,
+            t_or_d,
+            peptides,
+            score_result,
+            per_aa_score,
+            precursors,
+        ) in outputs:
+            for index, t_or_d, peptide, score, per_aa_scores, precursor in zip(
+                indexes,
+                t_or_d,
+                peptides,
+                score_result,
+                per_aa_score,
+                precursors,
+            ):
+                prec_charge = precursor[1]
+                prec_mz = precursor[2]
+                calc_mz = precursor[2]
+                self.out_writer.psms.append(
+                    (
+                        peptide,
+                        score,
+                        prec_charge,
+                        prec_mz,
+                        calc_mz,
+                        index,
+                        per_aa_scores,
+                        t_or_d,
+                    ),
+                )
+
+
+def _calc_match_score(
+    batch_all_aa_scores: torch.Tensor, truth_aa_indicies: torch.Tensor
+) -> List[float]:
+    """
+    Take in teacher-forced scoring of amino acids
+    of the peptides (in a batch) and use the truth labels
+    to calculate a score between the input spectra and
+    associated peptide. The score is the geometric
+    mean of the AA probabilities
+
+        Parameters
+        ----------
+        batch_all_aa_scores : torch.Tensor
+            Amino acid scores for all amino acids in
+            the vocabulary for every prediction made to generate
+            the associated peptide (for an entire batch)
+        truth_aa_indicies : torch.Tensor
+            Indicies of the score for each actual amino acid
+            in the peptide (for an entire batch)
+
+        Returns
+        -------
+        score : list[float], list[list[float]]
+            The score between the input spectra and associated peptide
+            (for an entire batch)
+            a list of lists of per amino acid scores
+            (for an entire batch)
+    """
+    # Remove trailing tokens from predictions,
+    batch_all_aa_scores = batch_all_aa_scores[:, :-1]
```

Comment:
Is this removing the score for the stop token? Or is there a prediction made after the stop token?

Reply:
If I remember correctly, this is removing the "start token", or something else not useful for score calculation.

Reply:
OK, just wanted to confirm that the score for the stop token is included.